File: app/src/main/java/com/arukai/uajpspeak/model/AudioPlayer.java
package com.arukai.uajpspeak.model;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.SoundPool;
import android.util.Log;

import static com.arukai.uajpspeak.activity.MainActivity.context;

public class AudioPlayer {
    private static final String LOG_TAG = "PLAYER";
    private static MediaPlayer mMediaPlayer = null;
    private static SoundPool soundPool = null;
    private static int streamId;

    private static void stopMP() {
        if (mMediaPlayer != null) {
            try {
                mMediaPlayer.reset();
                mMediaPlayer.release();
            } catch (Exception e){
                Log.e(LOG_TAG, "reset/release failed");
            }
            mMediaPlayer = null;
        }
    }

    private static void stopSP() {
        if (soundPool != null) {
            try {
                soundPool.stop(streamId);
                soundPool.release();
            } catch (Exception e){
                Log.e(LOG_TAG, "stop/release failed");
            }
            soundPool = null;
        }
    }

    public static void play(Context c, int rid) {
        stopMP();
        stopSP();

        mMediaPlayer = MediaPlayer.create(c, rid);
        mMediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
            @Override
            public void onCompletion(MediaPlayer mediaPlayer) {
                stopMP();
            }
        });

        try {
            mMediaPlayer.start();
        } catch (Exception e){
            Log.e(LOG_TAG, "start failed");
        }
    }

    public static void slowPlay(Context c, int rid) {
        stopMP();
        stopSP();

        final float playbackSpeed=0.75f;
        soundPool = new SoundPool(4, AudioManager.STREAM_MUSIC, 100);

        final int soundId = soundPool.load(c, rid, 1);
        AudioManager mgr = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        final float volume = mgr.getStreamMaxVolume(AudioManager.STREAM_MUSIC);

        soundPool.setOnLoadCompleteListener(new SoundPool.OnLoadCompleteListener()
        {
            @Override
            public void onLoadComplete(SoundPool arg0, int arg1, int arg2)
            {
                streamId = soundPool.play(soundId, volume, volume, 1, 0, playbackSpeed);
            }
        });
    }

}

File: app/src/main/java/com/arukai/uajpspeak/model/AudioPlayer.kt
package com.arukai.uajpspeak.model

import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.SoundPool
import android.util.Log
import com.arukai.uajpspeak.activity.MainActivity

object AudioPlayer {
    private const val LOG_TAG = "PLAYER"
    private var mMediaPlayer: MediaPlayer? = null
    private var soundPool: SoundPool? = null
    private var streamId: Int = 0

    private fun stopMP() {
        mMediaPlayer?.let {
            try {
                it.reset()
                it.release()
            } catch (e: Exception) {
                Log.e(LOG_TAG, "reset/release failed")
            }
            mMediaPlayer = null
        }
    }

    private fun stopSP() {
        soundPool?.let {
            try {
                it.stop(streamId)
                it.release()
            } catch (e: Exception) {
                Log.e(LOG_TAG, "stop/release failed")
            }
            soundPool = null
        }
    }

    fun play(c: Context, rid: Int) {
        stopMP()
        stopSP()

        mMediaPlayer = MediaPlayer.create(c, rid)
        mMediaPlayer?.setOnCompletionListener {
            stopMP()
        }

        try {
            mMediaPlayer?.start()
        } catch (e: Exception) {
            Log.e(LOG_TAG, "start failed")
        }
    }

    fun slowPlay(c: Context, rid: Int) {
        stopMP()
        stopSP()

        val playbackSpeed = 0.75f
        soundPool = SoundPool(4, AudioManager.STREAM_MUSIC, 100)

        val soundId = soundPool?.load(c, rid, 1) ?: return
        val mgr = MainActivity.context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val volume = mgr.getStreamMaxVolume(AudioManager.STREAM_MUSIC).toFloat()

        soundPool?.setOnLoadCompleteListener { _, _, _ ->
            streamId = soundPool?.play(soundId, volume, volume, 1, 0, playbackSpeed) ?: 0
        }
    }
}


